### PR TITLE
FEAT-#1944: Added ability to broadcast multi-partitioned frames

### LINKED ISSUE #1944: Add ability to of broadcasting sub-DataFrames which located in a many partitions
Currently, all partition broadcasting approaches at ModinFrame works only if frame at broadcasted axis located in a single partition.
For example, if we want to broadcast a multicolumn frame it's columns could be located in different partitions then we have no ways to do it (could be useful for example in a multicolumn group by).

### PR DESCRIPTION
**Pull request by dchigarev**
Signed-off-by: Dmitry Chigarev <dmitry.chigarev@intel.com>

## What do these changes do?

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves #1944 
- [x] tests added and passing

### CODE REVIEW

File: modin/engines/base/frame/partition_manager.py [hunk 1]
```diff
@@ -167,7 +152,7 @@ def get_partitions(index):
         return new_partitions
 
     @classmethod
-    def broadcast_apply(cls, axis, apply_func, left, right):
+    def broadcast_apply(cls, axis, apply_func, left, right, other_name="r"):
```
Comment (devin-petersohn) [inline modin/engines/base/frame/partition_manager.py:155]: Please add docstring label for `other_names`.

File: modin/engines/dask/pandas_on_dask/frame/partition_manager.py [hunk 2]
```diff
@@ -20,29 +20,34 @@
...
 from .partition import PandasOnDaskFramePartition
 from modin.error_message import ErrorMessage
+import pandas
 
 from distributed.client import _get_global_client
 import cloudpickle as pkl
 
 
-def deploy_func(df, other, apply_func, call_queue_df=None, call_queue_other=None):
+def deploy_func(df, apply_func, call_queue_df=None, call_queues_other=None, *others):
```
Comment (devin-petersohn) [inline modin/engines/dask/pandas_on_dask/frame/partition_manager.py:29]: What is the purpose of this change?

Comment (dchigarev) [reply modin/engines/dask/pandas_on_dask/frame/partition_manager.py:29]: `deploy_func` is a remote function, when it obtains future it will be materialized inside that function, so since we need to pass a variable number of `others` futures,  we're packing it in a `*others`. (if we pass a list of futures inside the function it will be still a list of unmaterialized futures)

**Review by YarShev — COMMENTED** [review]
@dchigarev , as I see, your added test falls down to here [place#1](https://github.com/modin-project/modin/blob/master/modin/pandas/dataframe.py#L468) and then here [place#2](https://github.com/modin-project/modin/blob/master/modin/engines/base/frame/partition_manager.py#L82) where the exception (`AttributeError: 'numpy.ndarray' object has no attribute 'drain_call_queue'`) is arisen. Because array `by_parts` can be 2D the exception is arisen in this case. Can we do just [these changes](https://github.com/modin-project/modin/pull/1945/files#diff-ba68222217d05b3a1a8da70a6059e392R170) and avoid the exception? Also, regarding to the place#1, which I mentioned. I think it worth deleting the warning because we will able to support multi-column groupby with these changes.

**Comment by dchigarev** [comment]
> @dchigarev , as I see, your added test falls down to here [place#1](https://github.com/modin-project/modin/blob/master/modin/pandas/dataframe.py#L468) and then here [place#2](https://github.com/modin-project/modin/blob/master/modin/engines/base/frame/partition_manager.py#L82) where the exception (`AttributeError: 'numpy.ndarray' object has no attribute 'drain_call_queue'`) is arisen. Because array `by_parts` can be 2D the exception is arisen in this case. Can we do just [these changes](https://github.com/modin-project/modin/pull/1945/files#diff-ba68222217d05b3a1a8da70a6059e392R170) and avoid the exception? Also, regarding to the place#1, which I mentioned. I think it worth deleting the warning because we will able to support multi-column groupby with these changes.

adding only changes that you mentioned apparently do not resolve the problem with broadcasting multi-partitioned frames. Yeah, we will avoid exception but still have no ability to broadcast that frame. For broadcasting in addition to those changes, we should contact columns to make from them a single dataframe.

**Comment by dchigarev** [comment]
@devin-petersohn here is the performance measurements for Ray engine

<details><summary>script</summary>


```python
import numpy as np
import os
import pandas
from itertools import permutations, repeat
from timeit import default_timer as timer

REPEAT_TIMES = 10

def get_permutations(l, r):
    return [(l_x, r_x) for l_x in l for r_x in r]


def generate_data_file(
    filename=None,
    nrows=100000,
    ncols=128,
    multiline_rows=False,
    force=False,
    rand_low=-100,
    rang_high=100,
    log=True,
):
    if filename is None:
        filename = os.path.abspath(
            f"int_dataset-{nrows},{ncols},{rand_low},{rang_high},{multiline_rows}.csv"
        )

    print("generating", filename)

    if os.path.exists(filename) and not force:
        return filename

    data = {
        f"col{i}": np.concatenate(
            [
                np.concatenate(
                    [
                        ["some\nvery very very\nlong string\nwith many multilines"],
                        np.random.randint(rand_low, rang_high, 9),
                    ]
                )
                for _ in np.arange(ncols // 10)
            ]
        )
        if (i % 10 == 0 and multiline_rows)
        else list(np.random.randint(rand_low, rang_high, nrows // 100)) * 100
        for i in np.arange(ncols)
    }
    print("dict generated!")
    df = pandas.DataFrame(data)
    print("dataframe created!")
    df.to_csv(filename)
    print("csv ready!")
    return filename


def groupby_single_column(df):
    grp = df.groupby(df.columns[0])
    repr(grp.sum())


def groupby_multiple_columns(df):
    grp = df.groupby(df.columns[:10])
    repr(grp.sum())


def binary_operation(df):
    repr(df + df[df.columns[1]])


if __name__ == "__main__":
    import modin.pandas as pd

    n_cores = [8, max(1, pd.DEFAULT_NPARTITIONS)]
    dimensions = [500, 50000, 500000, 2500000]
    test_functions = [
        groupby_single_column,
        groupby_multiple_columns,
        binary_operation,
    ]

    for dimension in dimensions:
        generate_data_file(nrows=dimension)
    print("ALL TEST DATA GENERATED!")

    test_cases = get_permutations(n_cores, dimensions)
    results = {
        key_core: {
            key_fname.__name__: {
                key_dimension: None for key_dimension in dimensions
            }
            for key_fname in test_functions
        }
        for key_core in n_cores
    }

    for cores, dim in test_cases:
        filename = generate_data_file(nrows=dim, force=False)
        pd.DEFAULT_NPARTITIONS = cores

        md_df = pd.read_csv(filename)
        print("test file readed")
        for fn in test_functions:
            result = float("inf")
            for _ in range(max(1, REPEAT_TIMES)):
                t1 = timer()
                fn(md_df)
                result = min(result, timer() - t1)

            results[cores][fn.__name__][dim] = result

    print(results)


```



</details>

**8 cores:**
| Data size | groupby on single column (#1945) | groupby on single column (master) | groupby on multiple columns (#1945) | groupby on multiple columns (master) | binary operation (#1945) | binary operation (master)
| --- | --- | --- | --- | --- | --- | --- |
|  500x128 | 0.21s | 0.20s | 0.35s | 0.34s | 0.19s | 0.17s
|  50000x128 | 0.37s | 0.37s | 0.61s | 0.61s | 0.22s | 0.22s
|  500000x128 | 2.17s | 2.33s | 3.55s | 3.60s | 0.46s | 0.43s
|  2500000x128 | 10.95s | 10.93s | 17.63s | 17.95s | 1.50s | 1.43s


**112 cores:**
| Data size | groupby on single column (#1945) | groupby on single column (master) | groupby on multiple columns (#1945) | groupby on multiple columns (master) | binary operation (#1945) | binary operation (master)
| --- | --- | --- | --- | --- | --- | --- |
|  500x128 | 1.22s | 1.21s | 1.88s | 1.87s | 1.10s | 1.10s
|  50000x128 | 1.47s | 1.46s | 2.38s | 2.38s | 0.82s | 0.84s
|  500000x128 | 2.95s | 2.97s | 7.65s | 7.50s | 1.02s | 1.06s
|  2500000x128 | 9.95s | 9.93s | 32.85s | 33.39s | 1.62s | 1.78s

As we can see, there is no performance impact in the common case (when we don't need to broadcast multi-partitioned frame)

**Comment by dchigarev** [comment]
@devin-petersohn 
> This is a pretty big change to internal APIs, there may be some ways to avoid all of the extra internal APIs.

What changes in internal APIs do you mean? What I see is that I've added an extra argument `other_name` to `broadcast_apply`, but since it has default value `"r"` the way of working with that function doesn't change

File: modin/engines/base/frame/partition_manager.py (hunk 1 above)
Comment (dchigarev) [reply modin/engines/base/frame/partition_manager.py:155]: done

**Comment by YarShev** [comment]
@dchigarev , @devin-petersohn , on the whole it looks pretty good, but I have the same question that I mentioned above about May be it's worth deleting [the warning](https://github.com/modin-project/modin/blob/master/modin/pandas/dataframe.py#L464)?

**Comment by devin-petersohn** [comment]
Yes @YarShev, I think to delete that warning would be good.

**Comment by dchigarev** [comment]
> Yes @YarShev, I think to delete that warning would be good.

@YarShev @devin-petersohn warning removed.

**Comment by dchigarev** [comment]
@YarShev @devin-petersohn just a reminder to review)

**Review by YarShev — APPROVED** [review]
@dchigarev , thanks, LGTM!
